identity/src/main/java/com/android/identity/util/ApplicationData.java

```diff
@@ -0,0 +1,163 @@
+     * @param value the value or {@code null} to remove.
+     */
+    void setApplicationData(@NonNull String key, @Nullable byte[] value);
+
+    /**
+     * Sets application specific data as a string.
+     *
+     * <p>Like {@link #setApplicationData(String, byte[])} but encodes the given value
+     * as an UTF-8 string before storing it.
+     *
+     * @param key the key for the data.
+     * @param value the value or {@code null} to remove.
+     */
+    void setApplicationData(@NonNull String key, @Nullable String value);
+
+    /**
+     * Sets application specific data as a {@link Number}.
+     *
+     * <p>Like {@link #setApplicationData(String, byte[])} but encodes the given value
+     * using <a href="http://cbor.io/">CBOR</a> before storing it.
+     *
+     * @param key the key for the data.
+     * @param value the value or {@code null} to remove.
+     */
+    void setApplicationData(@NonNull String key, @Nullable Long value);
+
+    /**
+     * Sets application specific data as a {@link Timestamp}.
+     *
+     * <p>Like {@link #setApplicationData(String, byte[])} but encodes the given value
+     * using <a href="http://cbor.io/">CBOR</a> before storing it.
+     *
+     * @param key the key for the data.
+     * @param value the value or {@code null} to remove.
+     */
+    void setApplicationData(@NonNull String key, @Nullable Timestamp value);
+
+    /**
+     * Sets application specific data as a boolean.
+     *
+     * <p>Like {@link #setApplicationData(String, byte[])} but encodes the given value
+     * using <a href="http://cbor.io/">CBOR</a> before storing it.
+     *
+     * @param key the key for the data.
+     * @param value the value or {@code null} to remove.
+     */
+    void setApplicationData(@NonNull String key, @Nullable Boolean value);
+
+    /**
+     * Gets application specific data.
+     *
+     * <p>Gets data previously stored with {@link #getApplicationData(String)}.
+     *
+     * @param key the key for the data.
+     * @return the value or {@code null} if there is no value for the given key.
+     */
+    @Nullable byte[] getApplicationData(@NonNull String key);
+
+    /**
+     * Gets application specific data as a string.
+     *
+     * <p>Like {@link #getApplicationData(String)} but decodes the value as
+     * an UTF-8 string before returning it.
+     *
+     * @param key the key for the data.
+     * @return the value or {@code null} if there is no value for the given key.
+     */
+    @Nullable String getApplicationDataString(@NonNull String key);
+
+    /**
+     * Gets application specific data as a {@link Number}.
+     *
+     * <p>Like {@link #getApplicationData(String)} but decodes the value as
+     * a {@link Long} before returning it.
+     *
+     * @param key the key for the data.
+     * @return the value or {@code null} if there is no value for the given key.
+     */
+    @Nullable Long getApplicationDataNumber(@NonNull String key);
```

Comment:
Instead of the boxed type, use `long` instead. If the data element doesn't exist, throw `IllegalArgumentException`. Applications that are unsure if a data element exist can use the generic `getApplicationData()` to check (it can return `null`). Same for getters of other primitive types.

The reason for this is that these functions are here for convenience. It would be annoying having to test for non-nullness and in a properly written application that will not happen (b/c the application is carefully written to have always `set()` whatever it expects to `get()` later) so it's fine to use exceptions for this.